### PR TITLE
[OP] Change fused matmul layout type and number thread for small size inputs.

### PR DESCRIPTION
**Pull request by marvin-Yu**
This PR mainly change the `_MklFuedMatMul` layout type, It deleted those unnecessary tensor format changes and reduce framework overhead.
- Before applying this PR.
image
- After applying this PR..
image

Performance changing
|_MklFusedMatMul performance|Time(ms)|percent|
|:--:|:--:|:--:|
|DeepRec FP32 - Before|8.862|baseline|
|DeepRec FP32 - After|8.689|101%|

### CODE REVIEW

File: tensorflow/core/kernels/mkl_matmul_op_fused.cc [hunk 1]
```diff
@@ -301,7 +295,7 @@ class MklFusedMatMulOp : public MklDnnMatMulOpBase<T, T> {
       Name("_MklFusedMatMul")                                  \
           .Device(DEVICE_CPU)                                  \
           .TypeConstraint<type>("T")                           \
-          .Label(mkl_op_registry::kMklLayoutDependentOpLabel), \
+          .Label(mkl_op_registry::kMklNameChangeOpLabel),      \
```
Comment (changqi1) [inline tensorflow/core/kernels/mkl_matmul_op_fused.cc:298]: magic name: kMklNameChangeOpLabel

Comment (changqi1) [reply tensorflow/core/kernels/mkl_matmul_op_fused.cc:298]: If you don't know how to define it, please give some annotation for it.

**Comment by changqi1** [comment]
For "[OP] Add env Marco for MKL ops inter/intra thread setting.", you only set intra thread number, not inter.

File: tensorflow/core/kernels/mkl_matmul_op_fused.cc (hunk 1 above)
Comment (marvin-Yu) [reply tensorflow/core/kernels/mkl_matmul_op_fused.cc:298]: Intel-tf designs these different kernel type, here have some explanation of  this two kind of kernel type.
``` c++
// MKL operators whose kernels are registered with 'MklLayoutDependentOp' label
// (e.g., MklConv2D) understand input tensors in MKL layout. These operators
// get additional meta-tensors for actual input tensors.
static const char* kMklLayoutDependentOpLabel = "MklLayoutDependentOp";
static const char* kMklLayoutDependentOpLabelPattern =
    "label='MklLayoutDependentOp'";
// MKL operators whose kernels are registered with 'MklNameChangeOp' label
// (e.g., MklMatMul, MklTranspose) do not understand input tensors in MKL
// layout. These operators do not get additional meta-tensors. The signatures of
// these operators are the same as the original TensorFlow operators that they
// correspond to. So these ops just go through a name change during graph
// rewrite pass.
static const char* kMklNameChangeOpLabel = "MklNameChangeOp";
static const char* kMklNameChangeOpLabelPattern = "label='MklNameChangeOp'";
static const char* kMklQuantizedOpLabel = "QuantizedMklOp";
static const char* kMklQuantizedOpLabelPattern = "label='QuantizedMklOp'";
```

Comment (changqi1) [reply tensorflow/core/kernels/mkl_matmul_op_fused.cc:298]: Sure!

**Comment by marvin-Yu** [comment]
>

**Comment by marvin-Yu** [comment]
> For "[OP] Add env Marco for MKL ops inter/intra thread setting.", you only set intra thread number, not inter.

Got it, will change the commit content later.

**Comment by marvin-Yu** [comment]
> 你在PR中提交的3个commit顺序是反的，需要正过来

Todo(marvin): change the commit sequence, avoid potential compilation issues

File: tensorflow/core/kernels/mkl_matmul_op.cc [hunk 2]
```diff
@@ -161,8 +161,19 @@ class MklMatMulOp : public OpKernel {
...
 #ifdef ENABLE_MKLDNN_THREADPOOL
     MklDnnThreadPool eigen_tp(ctx);
-    sgemm(char_transa, char_transb, m, n, k, alpha, a, lda, b, ldb, beta, c,
-          ldc, &eigen_tp);
+    // With threadpool , the runtime overhead is comparable to the kernel
+    // execution for small kernel sizes. For such sizes, it may be better to run
+    // the kernel single threaded. Here we are coming up with a cost model based
+    // on L1 sizes. If we find that matrices are small enough, we will execute
+    // single threaded. This may need tuning.
+    if (ExecuteSingleThreadedGemm(m, n, k)) {
```
Comment (changqi1) [inline tensorflow/core/kernels/mkl_matmul_op.cc:169]: if (ExecuteSingleThreadedGemm(m, n, k)) {
    MklDnnThreadPool eigen_tp(ctx, 1);
    sgemm(char_transa, char_transb, m, n, k, alpha, a, lda, b, ldb,beta, c, ldc, &eigen_tp);
} else {
    MklDnnThreadPool eigen_tp(ctx);
    sgemm(char_transa, char_transb, m, n, k, alpha, a, lda, b, ldb,beta, c, ldc, &eigen_tp);
}

Would use this ?

Comment (changqi1) [reply tensorflow/core/kernels/mkl_matmul_op.cc:169]: dnnl_sgemm(char_transa, char_transb, m, n, k, alpha, a, lda, b, ldb, beta, c, ldc); will use all cores to compute sgemm, like the following #else part.

**Pull request closed without merging**